tests/python/test_multidevice.py

```diff
@@ -112,3 +113,362 @@ def multidevice_schedule(self):
+            shape=[b, s, h],
+            contiguity=[True if b > 1 else None, True, True],
+            dtype=DataType.BFloat16,
+        )
+        self.layernorm0_weight = self.define_tensor(
+            shape=[h],
+            contiguity=[True],
+            dtype=DataType.BFloat16,
+        )
+        self.layernorm0_bias = self.define_tensor(
+            shape=[h],
+            contiguity=[True],
+            dtype=DataType.BFloat16,
+        )
+        self.mha_linear0_weight = self.define_tensor(
+            shape=[h * 3, h],
+            contiguity=[True, True],
+            dtype=DataType.BFloat16,
+        )
+        self.mha_linear0_bias = self.define_tensor(
+            shape=[h * 3],
+            contiguity=[True],
+            dtype=DataType.BFloat16,
+        )
+        self.mha_linear1_weight = self.define_tensor(
+            shape=[h, h],
+            contiguity=[True, True],
+            dtype=DataType.BFloat16,
+        )
+        self.mha_linear1_bias = self.define_tensor(
+            shape=[h],
+            contiguity=[True],
+            dtype=DataType.BFloat16,
+        )
+        self.layernorm1_weight = self.define_tensor(
+            shape=[h],
+            contiguity=[True],
+            dtype=DataType.BFloat16,
+        )
+        self.layernorm1_bias = self.define_tensor(
+            shape=[h],
+            contiguity=[True],
+            dtype=DataType.BFloat16,
+        )
+        self.mlp_linear0_weight = self.define_tensor(
+            shape=[d, h * 4 // d, h],
+            contiguity=[True if d > 1 else None, True, True],
+            dtype=DataType.BFloat16,
+        )
+        self.mlp_linear0_bias = self.define_tensor(
+            shape=[d, h * 4 // d],
+            contiguity=[True if d > 1 else None, True],
+            dtype=DataType.BFloat16,
+        )
+        self.mlp_linear1_weight = self.define_tensor(
+            shape=[d, h, h * 4 // d],
+            contiguity=[True if d > 1 else None, True, True],
+            dtype=DataType.BFloat16,
+        )
+        self.mlp_linear1_bias = self.define_tensor(
+            shape=[h],
+            contiguity=[True],
+            dtype=DataType.BFloat16,
+        )
+        S17 = self.define_scalar(0.00000, dtype=DataType.Double)
+        S18 = self.define_scalar(1.00000, dtype=DataType.Double)
+        T23 = self.ops.uniform(
+            S17,
+            S18,
+            shape=[b, s, h],
+            rng_seed=S1,
+            rng_offset=S0,
+            dtype=DataType.BFloat16,
+        )
+        S24 = self.define_scalar(0.00000, dtype=DataType.Double)
+        S25 = self.define_scalar(1.00000, dtype=DataType.Double)
+        T30 = self.ops.uniform(
+            S24,
+            S25,
+            shape=[b, s, h],
+            rng_seed=S3,
+            rng_offset=S2,
+            dtype=DataType.BFloat16,
+        )
+        T31 = self.ops.cast(self.input, dtype=DataType.Float)
+        S32 = self.define_scalar(0.900000, dtype=DataType.Double)
```

Comment:
Shouldn't we make `0.900000` a variable ?

Reply:
Good question here and several other questions about readability down below. 

As mentioned [here](https://github.com/NVIDIA/Fuser/pull/3045/files#diff-31dc564b0e6d2c5e614a5ee718088699ed239fb7cbe8c19a4bf7802ec355a7c8R118-R121), this fusion definition is a port from the nvFuser fusion definition generated by Thunder. So I was trying to balance between making the code readable (e.g. giving constants names and extracting common stuff into functions) and minimizing manual changes when next time I update the test due to changes in Thunder. 

I'm on the fence and would love to hear arguments either way.

Reply:
Ok, I did not realize that. I am personally in favor of making the code readable, because 1) we shouldn't expect fundamental change from thunder, and 2) imo we don't necessarily need to track in this unit test all small changes from thunder: this test is useful because it is _one_ implementation of transformer, and it's ok if it differs from the actual thunder generation.

Not having DropOut, LayerNorm, GeLu scooped out as auxiliary functions makes imo the code error-prone, not flexible, and not readable.

I would, at least, add comments delimiting the different primitives.

This being said, I might be missing the actual goal of this unit test, so feel free to do as you think is best

Reply:
Ok, I did not realize that. I am personally in favor of making the code readable, because 1) we shouldn't expect fundamental change from thunder, and 2) imo we don't necessarily need to track in this unit test all small changes from thunder: this test is useful because it is _one_ implementation of transformer, and it's ok if it differs from the actual thunder generation.

Not having DropOut, LayerNorm, GeLu scooped out as auxiliary functions makes imo the code error-prone, not flexible, and not readable.

I would, at least, add comments delimiting the different primitives.

This being said, I might be missing the actual goal of this unit test, so feel free to do as you think is best

Reply:
I agree with most of your points. I recommend we merge this and discuss the plan after @cowanmeg is back. While I love cleanups, I'll feel more comfortable doing so if I know they'll stay. 

The purpose of this test is for performance comparison. test_transformer_engine is the baseline, SOTA. This test is to reflect what Thunder/nvFuser will generate. "Will" because we've yet to integrate multi-GPU nvFuser into Thunder. For that, I'd like it to be as close as possible to what Thunder will generate. When the integration is ready, we can/should remove this test, and build and track the benchmark in Thunder instead. 

That said, there's a different purpose that you are probably alluding to: build a best performing sharded transformer layer using nvFuser's API, regardless of the framework. There's definitely value to that, and in that case I'm all for optimizing readability. However, there's DistributedTransformerTest.Forward/Backward that serves the same purpose. We should discuss whether/how the two should co-exist. 

An incidental note: due to #2962, it's currently hard to verify any cleanup is behavior neutral.

Reply:
Alright